### PR TITLE
dbcache

### PR DESCRIPTION
**Pull request by ovkulkarni**
This allows user information needed for performance to be cached in the database. We should run `./manage.py set_db_cache` once deployed to set the cache, and then per-user cache can be cleared using the GUI. Accessing the cache information should be done by using filter: `User.objects.filter(cache__gender=True)` or `get_or_set_cache`: `User.get_user(username="2019okulkarn").get_or_set_cache('grade_number')`. The first query would get all males, and the second query would get my grade number as an `int`.

### CODE REVIEW

**Comment by pefoley2** [comment]
Please fix the unnecessary parens and add a test.

File: intranet/apps/users/models.py [hunk 1]
```diff
@@ -433,7 +436,16 @@ def full_name(self):
         This is required for subclasses of User.
 
         """
-        return self.common_name
+
+        if self.get_from_cache('first_name') is None:
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: Please factor this logic out to a function to avoid duplication.

File: intranet/apps/users/models.py [hunk 2]
```diff
@@ -743,11 +783,41 @@ def birthday(self):
 
     @property
     def is_male(self):
-        return self.sex.lower()[:1] == "m" if self.sex else False
+        if self.get_from_cache('gender') is None:
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: Factor this out as well

File: intranet/apps/users/models.py [hunk 3]
```diff
@@ -1048,8 +1121,11 @@ def is_teacher(self):
             Boolean
 
         """
-
-        return self.user_type == "tjhsstTeacher"
+        if not self.get_from_cache('objectClass'):
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: And this

File: intranet/apps/users/models.py [hunk 4]
```diff
@@ -1554,6 +1644,65 @@ def clear_cache(self):
...
+            "first_name": "first_name",
+            "last_name": "last_name",
+            "graduation_year": "graduation_year"
+        }
+        try:
+            val = getattr(self.cache, attribute)
+            if not val:
+                raise AttributeError
+            return val
+        except AttributeError:
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: don't need the overhead of try-catch, just do something like:
if val: return val
do other stuff

File: intranet/apps/users/models.py [hunk 5]
```diff
@@ -1554,6 +1644,65 @@ def clear_cache(self):
...
+                self.cache = UserCache.objects.create()
+                self.cache.save()
+                self.save()
+            if attribute in mappings:
+                if attribute == "gender":
+                    if self.sex:
+                        self.cache.gender = True if self.sex.lower()[:1] == "m" else False
+                    else:
+                        return None
+                elif attribute == "graduation_year":
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: why not use isdigit instead of hard-coding?

**Comment by pefoley2** [comment]
Also please rebase :)

File: intranet/apps/users/models.py (hunk 5 above)
Comment (ovkulkarni) [inline intranet/apps/users/models.py]: where should I use isdigit?

File: intranet/apps/users/models.py [hunk 6]
```diff
@@ -1128,18 +1153,17 @@ def is_simple_user(self):
...
     def male(self):
         """Return if the user is male."""
-        return self.sex and self.sex == "M"
+        return self.is_female
 
     @property
     def female(self):
         """Return if the user is female."""
-        return self.sex and self.sex == "F"
+        return self.is_male
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: This needs to be inverted :)

File: intranet/apps/users/models.py [hunk 7]
```diff
@@ -1554,6 +1578,68 @@ def clear_cache(self):
...
             cache.delete(User.create_secure_cache_key(":".join((self.dn, attr))))
+        logger.debug("Clearing cached DB values")
+        if self.cache:
+            UserCache.objects.get(pk=self.cache.pk).delete()
+        self.set_cache()
+
+    def set_cache(self):
+        logger.debug("Setting DB cache using LDAP values")
+        if not self.cache:
+            self.cache = UserCache.objects.create(objectClass=self.user_type, first_name=self.first_name, last_name=self.last_name)
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: This would be more readable if line-wrapped

File: intranet/apps/users/models.py [hunk 8]
```diff
@@ -1554,6 +1578,68 @@ def clear_cache(self):
...
+                        setattr(self.cache, attribute, self.__getattr__(mappings[attribute]))
+                self.cache.save()
+                return self.get_or_set_cache(attribute)
+            elif attribute == "grade_number":
+                logger.debug("Setting grade_number from LDAp")
+                self.cache.grade_number = self.grade.number if self.grade else None
+                self.cache.save()
+                return self.get_or_set_cache(attribute)
+            return None
+        except UserCache.DoesNotExist:
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: The whole thing doesn't need to be in this try-catch block, just have the part that actually throws the exception

File: intranet/apps/users/tests.py [hunk 9]
```diff
@@ -17,3 +22,48 @@ def test_dynamic_groups(self):
...
+        response = self.client.get(reverse('user_profile'), {"clear_cache": 1})
+        self.assertEqual(response.status_code, 302)
+
+    def test_property_without_cache(self):
+        user = self.make_admin()
+        # delete cache object related to user
+        UserCache.objects.filter(user=user).delete()
+        # make sure correct exception is raised when accessing a nonexistent object
+        with self.assertRaises(UserCache.DoesNotExist):
+            user.cache.gender
```
Comment (pefoley2) [inline intranet/apps/users/tests.py]: Why not just add a self.assertTrue or something to suppress the warning?

File: intranet/apps/users/tests.py [hunk 10]
```diff
@@ -17,3 +22,48 @@ def test_dynamic_groups(self):
...
+        # make sure user.cache does not exist
+        with self.assertRaises(UserCache.DoesNotExist):
+            user.cache
+        # gender is not accessible, so returns None and sets cache
+        self.assertEqual(user.get_or_set_cache('gender'), None)
+        # make sure cache was set by get_or_set_cache
+        self.assertNotEqual(user.cache, None)
```
Comment (pefoley2) [inline intranet/apps/users/tests.py]: assertIsNotNone

File: intranet/apps/users/tests.py [hunk 11]
```diff
@@ -17,3 +22,48 @@ def test_dynamic_groups(self):
...
+        # delete cache object related to user
+        UserCache.objects.filter(user=user).delete()
+        # make sure user.is_male is false and does not fail, i.e checks LDAP and sets cache
+        self.assertEqual(user.is_male, False)
```
Comment (pefoley2) [inline intranet/apps/users/tests.py]: assertFalse

File: intranet/templates/eighth/admin/delinquent_students.html [hunk 12]
```diff
@@ -70,7 +71,7 @@
                     <tr>
                         <td>
                             <a href="{% url 'user_profile' delinquent.user.id %}">
-                                {{ delinquent.user.last_name }}, {{ delinquent.user.first_name }}
+                                {{ delinquent.user.last_name' }}, {{ delinquent.user|from_cache:'first_name }}
```
Comment (pefoley2) [inline intranet/templates/eighth/admin/delinquent_students.html]: Remove stray '

File: intranet/templates/eighth/admin/out_of_building_schedules.html [hunk 13]
```diff
@@ -1,5 +1,6 @@
 {% extends "eighth/admin/eighth_admin_page_base.html" %}
 {% load static %}
+{% load cache %}
```
Comment (pefoley2) [inline intranet/templates/eighth/admin/out_of_building_schedules.html]: Do we actually need this in all files?

File: intranet/templates/eighth/profile_header.html [hunk 14]
```diff
@@ -20,7 +21,7 @@ <h2 title="{{ profile_user.ion_username }} ({{ profile_user.ion_id }})">
                 {% if profile_user.is_student %}
                     {% if profile_user.graduation_year %}
                         {{ profile_user.grade|capfirst }}
-                        (Grade {{ profile_user.grade.number }}: {{ profile_user.graduation_year }})
+                        (Grade {{ profile_user.grade_number' }}: {{ profile_user:from_cache('graduation_year) }})
```
Comment (pefoley2) [inline intranet/templates/eighth/profile_header.html]: Fix

File: intranet/apps/templatetags/cache.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
Comment (pefoley2) [inline intranet/apps/templatetags/cache.py]: Do we still need this at all?

File: intranet/apps/users/models.py [hunk 16]
```diff
@@ -743,11 +771,11 @@ def birthday(self):
 
     @property
     def is_male(self):
-        return self.sex.lower()[:1] == "m" if self.sex else False
+        return self.get_or_set_cache('gender') is True
 
     @property
     def is_female(self):
-        return self.sex.lower()[:1] == "f" if self.sex else False
+        return self.get_or_set_cache('gender') is False
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: Just do return not

File: intranet/apps/users/models.py [hunk 17]
```diff
@@ -743,11 +771,11 @@ def birthday(self):
 
     @property
     def is_male(self):
-        return self.sex.lower()[:1] == "m" if self.sex else False
+        return self.get_or_set_cache('gender') is True
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: is True is redundant

File: intranet/apps/users/models.py (hunk 8 above)
Comment (pefoley2) [inline intranet/apps/users/models.py]: What places? The only one i'm seeing that it should be expected is 1610, which could also be refactored to getattr + if, though that's not a big deal

File: intranet/apps/users/models.py [hunk 18]
```diff
@@ -1554,6 +1578,70 @@ def clear_cache(self):
...
+                else:
+                    logger.debug("Setting {} from LDAP".format(attribute))
+                    if str(self.__getattr__(mappings[attribute])).isdigit():
+                        setattr(self.cache, attribute, int(self.__getattr__(mappings[attribute])))
+                    else:
+                        setattr(self.cache, attribute, self.__getattr__(mappings[attribute]))
+                self.cache.save()
+                return self.get_or_set_cache(attribute)
+            elif attribute == "grade_number":
+                logger.debug("Setting grade_number from LDAp")
```
Comment (pefoley2) [inline intranet/apps/users/models.py]: spelling

File: intranet/templates/dashboard/seniors.html [hunk 19]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (pefoley2) [inline intranet/templates/dashboard/seniors.html]: Please remove unrelated whitespace changes

File: intranet/apps/users/models.py (hunk 16 above)
Comment (ovkulkarni) [inline intranet/apps/users/models.py]: that would give is_female = True even if they have no gender

File: intranet/apps/users/models.py (hunk 17 above)
Comment (ovkulkarni) [inline intranet/apps/users/models.py]: see the is_female reply, I'm going to keep it for consistency

File: intranet/apps/polls/views.py [hunk 20]
```diff
@@ -165,7 +167,7 @@ def fmt(num):
 def perc(num, den):
     if den == 0:
         return 0
-    return int(10000 * num / den) / 100
+    return round(num / den * 100, 2)
```
Comment (pefoley2) [inline intranet/apps/polls/views.py]: Don't you need to use 100.0 to ensure floating point arith is used?

File: intranet/apps/polls/views.py [hunk 21]
```diff
@@ -247,81 +253,47 @@ def handle_sap(q):
...
+    choices = []
+    for c in q.choice_set.all().order_by("num"):
+        votes = question_votes.filter(choice=c)
+        choices.append(generate_choice(c, votes, total_count, do_gender, show_answers))
+
+    """ Clear vote """
+    votes = question_votes.filter(clear_vote=True)
+    choices.append(generate_choice("Clear vote", votes, total_count, do_gender, show_answers))
+
+    """ Total """
```
Comment (pefoley2) [inline intranet/apps/polls/views.py]: Please use ### for this.
